encoders/firebase-decoders-json/src/json/java/com/google/firebase/decoders/json/JsonDataDecoderBuilderContext.java

```diff
@@ -0,0 +1,119 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.firebase.decoders.json;
+
+import android.util.JsonReader;
+
+import androidx.annotation.NonNull;
+
+import com.google.firebase.decoders.DataDecoder;
+import com.google.firebase.decoders.FieldRef;
+import com.google.firebase.decoders.ObjectDecoder;
+import com.google.firebase.decoders.TypeCreator;
+import com.google.firebase.decoders.TypeToken;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.util.HashMap;
+import java.util.Map;
+//TODO: remove Exception from method header
+//TODO: surpass warnings
+//TODO: Java Docs
+public class JsonDataDecoderBuilderContext implements DataDecoder {
+  private Map<TypeToken.ClassToken<?>, ObjectDecoderContextImpl<?>> objectDecodersCtx = new HashMap<>();
+  private Map<TypeToken.ClassToken<?>, TypeCreator<?>> typeCreators = new HashMap<>();
+  private JsonReader reader;
+
+  public JsonDataDecoderBuilderContext(@NonNull Map<TypeToken.ClassToken<?>, ObjectDecoder<?>> objectDecoders) {
+    for (Map.Entry<TypeToken.ClassToken<?>, ObjectDecoder<?>> entry : objectDecoders.entrySet()) {
+      TypeToken.ClassToken<?> classToken = entry.getKey();
+      ObjectDecoder objectDecoder = entry.getValue();
+      ObjectDecoderContextImpl objectDecoderCtx = ObjectDecoderContextImpl.of(classToken);
+      objectDecoder.decode(objectDecoderCtx);
+      TypeCreator<?> creator = objectDecoder.decode(objectDecoderCtx);
+      objectDecodersCtx.put(classToken, objectDecoderCtx);
+      typeCreators.put(classToken, creator);
+    }
+  }
+
+  @NonNull
+  @Override
+  public <T> T decode(@NonNull InputStream input, @NonNull TypeToken<T> typeToken) throws IOException {
+    this.reader = new JsonReader(new InputStreamReader(input, "UTF-8"));
+    return decode(typeToken);
+  }
+
+  private <T> T decode(TypeToken<T> typeToken) throws IOException {
+    if (typeToken instanceof TypeToken.ClassToken) {
+      TypeToken.ClassToken classToken = (TypeToken.ClassToken<T>) typeToken;
+      T val = (T) decodeObject(classToken);
+      return val;
+    } else if (typeToken instanceof TypeToken.ArrayToken) {
+      //TODO: Change typeParameter T in ArrayToken<T> to represent component type.
+      /**
+       * reader.beginArray();
+       * List<T> l = new LinkedList<>();
+       * while reader.hasNext:
+       *  T val = decode(arrayToken.getComponentType());
+       *  l.add(val);
+       * reader.endArray();
+       * return l.toArray();
+       */
+    }
+    return null;
+  }
+
+  private <T> T decodeObject(TypeToken.ClassToken<T> classToken) throws IOException {
+    if(classToken.getRawType().isPrimitive() || classToken.getRawType().equals(String.class))
+      return decodePrimitives(classToken.getRawType());
```

Comment:
nit: at some later point we should think of how to avoid auto-boxing for primitives, i.e. decodeObject returns `T` which will automatically auto-box any primitive type like `int` in a `Integer` only to unpack it back when it's injected into the class that is being decoded. So this is an optimization opportunity that can be saved for later.